src/pages/Chat/Chat/index.tsx

```diff
@@ -38,30 +38,68 @@ export default (): JSX.Element => {
   const [messages, setMessages] = useState<IMessage[]>([]);
   const [isLoadingEarlier, setIsLoadingEarlier] = useState<boolean>(false);
   const [chatRef, setChatRef] = useState<DocumentRefData>();
+  const [otherIsTyping, setOtherIsTyping] = useState<boolean>(false);
+  const typingTimeoutFunction = useRef<NodeJS.Timeout>();
+  const ignoreFirstOnInputTextChanged = useRef<boolean>(true);
+  const selfIsTyping = useRef<boolean>(false);
+  const firstLoad = useRef<boolean>(true);
 
   const sorter = (a: IMessage, b: IMessage): number => (b.createdAt as Date).getTime() - (a.createdAt as Date).getTime();
 
+  const naiveUnique = (msgs: IMessage[]): IMessage[] => {
+    // eslint-disable-next-line no-underscore-dangle
+    const messagesId = msgs.map((m) => m._id);
+
+    if (!messagesId.every((mid, i) => messagesId.indexOf(mid) === i)) {
+      // eslint-disable-next-line no-console
+      console.warn('Not unique array: autofixing but something is wrong');
+    }
+    // eslint-disable-next-line no-underscore-dangle
+    return msgs.filter((m, i) => messagesId.indexOf(m._id) === i);
+  };
+
+  const addMessages = useCallback((loadedMessages: QuerySnapshotDocuments): void => {
+    const chatMessages = loadedMessages.map((message) => {
+      const messageData = message.data();
+      return {
+        _id: message.id,
+        text: messageData.text,
+        createdAt: messageData.timestamp ? messageData.timestamp.toDate() : new Date(),
+        user: {
+          _id: messageData.sender.id,
+        },
+      };
+    });
+
+    setMessages((previousMessages) => naiveUnique(GiftedChat.append(previousMessages, chatMessages).sort(sorter)));
+  }, []);
+
+  const updateMessages = useCallback((): void => {
+    if (firstLoad.current) {
+      firstLoad.current = false;
+      return;
+    }
+    if (!chatRef) {
+      return;
+    }
+
+    const newestDate = messages[0]?.createdAt as Date;
+    if (newestDate) {
+      chatAPI.loadNewMessages(chatRef, newestDate).then((loadedMessages) => {
+        const filteredLoadedMessages = loadedMessages.docs.filter((doc) => doc.data().sender.id !== currentUserRef.id);
+        addMessages(filteredLoadedMessages);
+      });
+    }
+  }, [messages, chatRef, currentUserRef, addMessages]);
```

Comment:
If `messages` is a dependency of this callback and this callback is a dependency of the useEffect that subscribes the chat to onSnapshot changes of the chat document, doesn't this mean that any change in messages will trigger a re-subscription? If so, we would need to find a way to use newestDate while avoiding the useEffect re-triggering.

I actually had a crazy idea to put the loadMessages and subscription to the updateMessages in the same useEffect inside a `if(chatRef !== undefined)`. Not sure if it would work, but it could simplify the code by removing the need to check if(!chatRef) inside of loadMessages and updateMessages.

Reply:
A lot changed